polarCoordinateTraversal.m

```diff
@@ -111,75 +111,134 @@
 end
 
 % Find the intersection times for the ray and the radial shell containing the ray
-% at t_begin.
+% at t_begin; in order to determine if ray intersects grid.
 ray_unit_vector = 1 / sqrt(ray_direction(1)^2 + ray_direction(2)^2)...
     .* [ray_direction(1);  ray_direction(2)]';
 v = dot(ray_circle_vector,ray_unit_vector);
 discr = r^2 - (dot(ray_circle_vector,ray_circle_vector) - v^2);
 d = sqrt(discr);
-p1 = ray_origin + (v-d).*ray_unit_vector;
-p2 = ray_origin + (v+d).*ray_unit_vector;
-t1 = (p1(1)-ray_origin(1))/ray_direction(1);
-t2 = (p2(1)-ray_origin(2))/ray_direction(2);
+pa = ray_origin + (v-d).*ray_unit_vector;
+pb = ray_origin + (v+d).*ray_unit_vector;
+tol = 10^-16;
+if (ray_direction(1) < tol)
+    t1 = (pa(2) - ray_origin(2))/ray_direction(2);
+    t2 = (pb(2) - ray_origin(2))/ray_direction(2);
+else
+    t1 = (pa(1) - ray_origin(1))/ray_direction(1);
+    t2 = (pb(1) - ray_origin(1))/ray_direction(1);
+end
 
 % The ray may not intersect the grid at all. 
 % In particular, if the ray is outside the grid at t_begin.
-if (t1 < t_begin && t2 < t_begin )
+if t1 < t_begin && t2 < t_begin 
+    if verbose
+        fprintf("\nRay does not intersect polar grid for t_begin.")
+    end
+    return;
+end
+
+% It may be a tangent hit
+ if abs(t1 - t2) < tol
     if verbose
-        fprintf("\nRay does not intersect polar grid for t_begin.");
+        fprintf("\nTangent hit.")
     end
     return;
 end
 
 % If there is a ray/shell intersection, then set the radial voxel ID. 
 current_voxel_ID_r = 1 + (circle_max_radius - r)/delta_radius;
 if verbose
-    fprintf('RADIAL HIT INITIALIZED.\n');
+    fprintf('RADIAL HIT INITIALIZED.\n')
 end
 
-% Calculate Voxel ID Theta.
-current_voxel_ID_theta = floor(atan2(ray_start_y - circle_center_y, ray_start_x - circle_center_x) * num_angular_sections / (2 * pi));
+% II. Calculate Voxel ID Theta.
+tol = 10^-16;
```

Comment:
Do you not agree we should have a single ```tol``` variable that we pass into other function if they need it?

Reply:
yes, we should do that.